### PR TITLE
Add the FTPVL library along with unit tests

### PR DESCRIPTION
**Pull request by TypingKoala**
This PR includes the source for FTPVL (evaluation, fetchers, helpers, processors, styles, visualizers) along with corresponding pytest unit tests that it passes.

@HackerFoo

### CODE REVIEW

File: tests/test_style.py [hunk 1]
```diff
@@ -0,0 +1,53 @@
+""" Implement testes for Style class """
```
Comment (HackerFoo) [inline tests/test_style.py:1]: Typo

File: tests/test_visualizer.py [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+        styled_eval = eval1.process([ColorMapStyle(cmap)])
+        vis = SingleTableVisualizer(eval1, styled_eval, column_order=["a", "b", "c"])
+        vis.get_visualization() # should not fail
+
+        with pytest.raises(KeyError):
+            vis = SingleTableVisualizer(eval1, styled_eval, column_order=["d"])
+            vis.get_visualization()
+
+    def test_twotablevisualizer(self):
+        pass
```
Comment (HackerFoo) [inline tests/test_visualizer.py:58]: Why isn't this implemented? Please add a comment explaining why, or an issue number.

File: ftpvl/processors.py [hunk 3]
```diff
@@ -0,0 +1,420 @@
...
+                input_eval.get_df()
+                .sort_values(by=self._sort_col_names, ascending=self._reverse_sort)
+                .drop_duplicates(subset=self._duplicate_col_names)
+            )
+            return Evaluation(new_df, input_eval.get_eval_id())
+
+
+class AddNormalizedColumn(Processor):
+    """
+    Processor that groups rows by a column, calculates the maximum of the
```
Comment (HackerFoo) [inline ftpvl/processors.py:117]: For runtime, it makes more sense to normalize based on the minimum. Generally the comparison is made to the "best" measurement.

File: tests/test_visualizer.py (hunk 2 above)
Comment (TypingKoala) [reply tests/test_visualizer.py]: Resolved by removing references to non-existent TwoTableVisualizer.

File: ftpvl/processors.py (hunk 3 above)
Comment (TypingKoala) [reply ftpvl/processors.py]: Added support for direction to AddNormalizedColumn. Can specify to use either min or max for the baseline when normalizing.

File: docs/ip [hunk 4]
```diff
@@ -0,0 +1 @@
+52.249.181.108
```
Comment (HackerFoo) [inline docs/ip:1]: Why is this here?

File: tests/test_visualizer.py [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+
+class TestVisualizer():
+    """
+    Testing by partition.
+
+    DebugVisualizer()
+
+    SingleTableVisualizer()
+    """
+    def test_debugvisualizer(self):
```
Comment (macieksmuga) [inline tests/test_visualizer.py:19]: This is a good start. Let's discuss more robust testing of visualizations in the near future. It's literally one of the biggest problems I deal with, and there are various competing ways to do this.

File: ftpvl/helpers.py [hunk 6]
```diff
@@ -0,0 +1,160 @@
...
+    Returns
+    -------
+    dict
+        a copy of the input dictionary only containing keys starting with
+        `versions.`
+    """
+    return {k: v for k, v in obj.items() if k.startswith("versions.")}
+
+
+def rescale_actual_freq(freq: Union[int, float]) -> Union[int, float]:
```
[the remark this answers is not in the document]
Comment (TypingKoala) [reply ftpvl/helpers.py:64]: Right. The issue that this resolves is that we don't get any units from the test cases. Some of them return things like 82.01 and others are 82010000 without any meaningful differentiation. This is a bit of a hack that resolves it for now, but optimally we fix that in fpga-tool-perf.

File: ftpvl/fetchers.py [hunk 7]
```diff
@@ -0,0 +1,206 @@
...
+                    f"Unable to get build {build_num}. It might have failed.",
+                )
+                continue
+            try:
+                data += [resp.json()]
+            except json.decoder.JSONDecodeError:
+                print("Warning:", f"Unable to decode build {build_num}")
+
+        if len(data) == 0:
+            raise ValueError(f"All builds from eval_num {self.eval_num} failed")
```
[the remark this answers is not in the document]
Comment (TypingKoala) [reply ftpvl/fetchers.py:136]: I've improved the wording of the error message to: `"Unable to get any successful builds from eval_num {self.eval_num}."`. The goal is for the user to check hydra to ensure that the selected evaluation has at least one passing test case. Hopefully that fixes it.

File: ftpvl/fetchers.py [hunk 8]
```diff
@@ -0,0 +1,206 @@
...
+            Raised if all builds in a given eval failed.
+        """
+        # get build numbers from eval_num
+        resp = requests.get(
+            f"https://hydra.vtr.tools/jobset/{self.project}/{self.jobset}/evals",
+            headers={"Content-Type": "application/json"},
+        )
+        if resp.status_code != 200:
+            raise ConnectionError("Unable to get evals from server.")
+        evals_json = resp.json()
```
[the remark this answers is not in the document]
Comment (TypingKoala) [reply ftpvl/fetchers.py:112]: The json structure is not well documented as far as I'm aware, and is based on fpga-tool-perf. To rectify this, the fetcher preprocessing will raise a keyerror if the user specifies a field in the mapping parameter and an expected field is not present.

```python
for in_col_name, out_col_name in self.mapping.items():
    processed_row[out_col_name] = row[in_col_name]
```

File: ftpvl/fetchers.py [hunk 9]
```diff
@@ -0,0 +1,206 @@
...
+    def _preprocess(self, path: str) -> pd.DataFrame:
+        """
+        Given the path, load the data in pandas, process the data, and return
+        the resulting dataframe.
+        """
+        loaded_df = pd.read_json(path)
+        if self.mapping is None:
+            return loaded_df
+        else:
+            return loaded_df.filter(items=self.mapping.keys()).rename(
```
[the remark this answers is not in the document]
Comment (TypingKoala) [reply ftpvl/fetchers.py:201]: The assumption is that JSON dataframes are already preprocessed to ensure there is only the actual clock. This worked for the JSON dataframes I've seen so far.

File: ftpvl/helpers.py [hunk 10]
```diff
@@ -0,0 +1,160 @@
+""" Defines helper functions useful for classes """
+from typing import Union, Any
+
+from matplotlib.colors import Colormap
+
+
+def flatten(input_dict: dict) -> dict:
```
[the remark this answers is not in the document]
Comment (TypingKoala) [reply ftpvl/helpers.py:8]: This has been added

File: ftpvl/processors.py [hunk 11]
```diff
@@ -0,0 +1,435 @@
...
+    groupby : str
+        the column to group by
+    
+    input_col_name : str
+        the input column to normalize
+
+    output_col_name : str
+        the column to write the normalized values to
+
+    direction : int (either 1 or -1)
```
[the remark this answers is not in the document]
Comment (TypingKoala) [reply ftpvl/processors.py:136]: this will require a lot of changes to docs and implementation. Moving to issue #11

File: ftpvl/processors.py [hunk 12]
```diff
@@ -0,0 +1,435 @@
...
+    idx_value : str, optional
+        the value of the baseline result at idx_name, by default "vivado"
+    """
+
+    def __init__(
+        self,
+        normalize_direction: dict,
+        group_by: str = "project",
+        idx_name: str = "synthesis_tool",
+        idx_value: str = "vivado",
```
[the remark this answers is not in the document]
Comment (TypingKoala) [reply ftpvl/processors.py]: resolved by removing defaults

File: ftpvl/processors.py [hunk 13]
```diff
@@ -0,0 +1,435 @@
...
+    Parameters
+    ----------
+    input_col_name : str
+        the column name to map from
+
+    output_col_names : List[str]
+        the column names to map to
+
+    mapping : dict
+        a dictionary mapping a value to a list of values
```
[the remark this answers is not in the document]
Comment (TypingKoala) [reply ftpvl/processors.py]: this has been clarified. There is probably a more succinct way to explain it.

```
mapping : Dict[str, List[str]]
        a dictionary mapping a value to a list of values. The processor will
        look at the value at input_col_name, use it as the key to index into mapping, and 
        get the corresponding list of strings that will used as the values for the new 
        metrics with names in output_col_names.
```

File: ftpvl/processors.py [hunk 14]
```diff
@@ -0,0 +1,435 @@
...
+    Processor that reassigns current columns as indices for easier
+    visualization.
+
+    Reindexing is useful for grouping similar results in the final
+    visualization.
+
+    Parameters
+    ----------
+    reindex_names : List[str]
+        A list of column names to reindex
```
[the remark this answers is not in the document]
Comment (TypingKoala) [reply ftpvl/processors.py:237]: I improved the description of Reindex:

```
Reindexing is useful for grouping similar results in the final
    visualization, since each row is identified by a (usually-unique)
    value in the index. Indices are always the first columns, which
    improves readability of the final visualization.
```

The columns "Project", "synthesis_tool", "pr_tool", and "toolchain" are the indices in [this visualization](https://ftpvl.readthedocs.io/en/latest/intro/firststeps.html#visualization).

File: ftpvl/fetchers.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (macieksmuga) [reply ftpvl/fetchers.py:112]: Ok, thanks! I'll circle back to this, but it's not a blocker for this PR.

File: ftpvl/processors.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (macieksmuga) [reply ftpvl/processors.py:136]: Thanks!

File: tests/test_visualizer.py (hunk 5 above)
Comment (macieksmuga) [reply tests/test_visualizer.py:19]: (this isn't a blocker, I'm resolving this for now)

File: ftpvl/fetchers.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (macieksmuga) [reply ftpvl/fetchers.py:201]: Ok, thanks!

File: ftpvl/helpers.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (macieksmuga) [reply ftpvl/helpers.py:65]: Ok, noted. I'd double-check to see if there is some implicit unit convention within fpga-tool-perf, perhaps related to the exact clock symbol employed.

File: ftpvl/processors.py [hunk 15]
```diff
@@ -0,0 +1,435 @@
...
+        input_df = input_eval.get_df()
+
+        # if int, we might need to convert to float first
+        # (e.g. int(float("6.0")))
+        if int in self.types.values():
+            # create dictionary replacing int with float
+            pre_df_types = {
+                k: (v if v != int else float) for k, v in self.types.items()
+            }
+            input_df = input_df.astype(pre_df_types)
```
[the remark this answers is not in the document]
Comment (TypingKoala) [reply ftpvl/processors.py:62]: From the [`.astype()` source](https://github.com/pandas-dev/pandas/blob/v1.1.0/pandas/core/generic.py#L5384-L5547), it looks like the only error explicitly raised is a `KeyError`. I'll add a handler for that.

**Comment by macieksmuga** [comment]
Looks good to me!
